File: banchus-app/src/main/java/pe/nanamochi/banchus/protocol/AbstractPacketHandler.java
package pe.nanamochi.banchus.protocol;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.lang.reflect.ParameterizedType;
import pe.nanamochi.banchus.core.Packet;
import pe.nanamochi.banchus.core.Packets;
import pe.nanamochi.banchus.database.entity.Session;

public abstract class AbstractPacketHandler<T extends Packet> {
  private final Packets packetType;
  private final Class<T> packetClass;

  @SuppressWarnings("unchecked")
  public AbstractPacketHandler() {
    HandleClientPacket handleAnno = this.getClass().getAnnotation(HandleClientPacket.class);
    this.packetType = handleAnno.value();

    ParameterizedType superClass = (ParameterizedType) getClass().getGenericSuperclass();
    this.packetClass = (Class<T>) superClass.getActualTypeArguments()[0];
  }

  public final Packets getPacketType() {
    return packetType;
  }

  public final Class<T> getPacketClass() {
    return packetClass;
  }

  public abstract void handle(T packet, Session session, ByteArrayOutputStream responseStream)
      throws IOException;
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/infrastructure/protocol/AbstractPacketHandler.kt
package pe.nanamochi.banchus.infrastructure.protocol

import java.io.ByteArrayOutputStream
import pe.nanamochi.banchus.core.ClientPacket
import pe.nanamochi.banchus.core.PacketType
import pe.nanamochi.banchus.redis.entity.Session

abstract class AbstractPacketHandler<T : ClientPacket>(val type: PacketType) {
    abstract fun handle(packet: T, session: Session, responseStream: ByteArrayOutputStream)
}
